lib/modem_info/modem_info.c

```diff
@@ -0,0 +1,412 @@
+	.data_type = MODEM_DATA_TYPE_SHORT,
+};
+
+static const struct modem_info_data temp_data = {
+	.cmd = AT_CMD_TEMP,
+	.param_index = TEMP_PARAM_INDEX,
+	.param_count = TEMP_PARAM_COUNT,
+	.data_type = MODEM_DATA_TYPE_SHORT,
+};
+
+static const struct modem_info_data fw_data = {
+	.cmd = AT_CMD_FW_VERSION,
+	.param_index = FW_PARAM_INDEX,
+	.param_count = FW_PARAM_COUNT,
+	.data_type = MODEM_DATA_TYPE_STRING,
+};
+
+static const struct modem_info_data *const modem_data[] = {
+	[MODEM_INFO_RSRP] = &rsrp_data,
+	[MODEM_INFO_BAND] = &band_data,
+	[MODEM_INFO_MODE] = &mode_data,
+	[MODEM_INFO_OPERATOR] = &operator_data,
+	[MODEM_INFO_CELLID] = &cellid_data,
+	[MODEM_INFO_IP_ADDRESS] = &ip_data,
+	[MODEM_INFO_UICC] = &uicc_data,
+	[MODEM_INFO_BATTERY] = &battery_data,
+	[MODEM_INFO_TEMP] = &temp_data,
+	[MODEM_INFO_FW_VERSION] = &fw_data,
+};
+
+static const char *const modem_data_name[] = {
+	[MODEM_INFO_RSRP] = "RSRP",
+	[MODEM_INFO_BAND] = "BAND",
+	[MODEM_INFO_MODE] = "MODE",
+	[MODEM_INFO_OPERATOR] = "OPERATOR",
+	[MODEM_INFO_CELLID] = "CELLID",
+	[MODEM_INFO_IP_ADDRESS] = "IP ADDRESS",
+	[MODEM_INFO_UICC] = "SIM",
+	[MODEM_INFO_BATTERY] = "BATTERY",
+	[MODEM_INFO_TEMP] = "TEMP",
+	[MODEM_INFO_FW_VERSION] = "FW",
+};
+
+static rsrp_cb_t modem_info_rsrp_cb;
+
+static struct at_param_list m_param_list;
+static struct k_thread socket_thread;
+static K_THREAD_STACK_DEFINE(socket_thread_stack, THREAD_STACK_SIZE);
+static struct k_mutex socket_mutex;
+static int at_socket_fd = INVALID_DESCRIPTOR;
+static struct pollfd fds;
+static int nfds;
+
+static int at_cmd_send(int fd, const char *cmd, size_t size, char *resp_buffer)
+{
+	int len;
+
+	k_mutex_lock(&socket_mutex, K_FOREVER);
+
+	LOG_DBG("send: %s\n", cmd);
+	len = send(fd, cmd, size, 0);
+	LOG_DBG("len: %d\n", len);
+	if (len != size) {
+		LOG_DBG("send: failed\n");
+		return -EIO;
+	}
+
+	if (resp_buffer != NULL) {
+		len = recv(fd, resp_buffer, CONFIG_MODEM_INFO_BUFFER_SIZE, 0);
+	}
+
+	if (len < 0) {
+		return -EIO;
+	}
+
+	k_mutex_unlock(&socket_mutex);
+
+	return 0;
+}
+
+static bool is_cesq_notification(char *buf, size_t len)
+{
+	if ((len < CMD_SIZE(AT_CMD_CESQ_RESP)) ||
+	    (buf[4] != AT_CMD_CESQ_RESP[4])) {
+		return false;
+	}
+
+	return strstr(buf, AT_CMD_CESQ_RESP) ? true : false;
+}
+
+enum modem_data_type modem_info_type_get(enum modem_info info)
+{
+	__ASSERT(info < MODEM_INFO_COUNT, "Invalid argument.");
+
+	return modem_data[info]->data_type;
+}
+
+int modem_info_name_get(enum modem_info info, char *name)
+{
+	int len;
+
+	if (name == NULL) {
+		return -EINVAL;
+	}
+
+	len = strlen(modem_data_name[info]);
+
+	memcpy(name,
+		modem_data_name[info],
+		len);
+
+	return len <= 0 ? -EINVAL : len;
+}
+
+int modem_info_short_get(char *out, size_t param_index,
+			size_t len, size_t off)
+{
+	int err;
+	u16_t param_value;
+
+	err = at_params_get_short(&m_param_list, param_index,
+				  &param_value);
+
+	if (err != 0) {
+		return err;
+	}
+
+	return snprintf(out, len,
```

Comment:
Is this the place to turn the value back into a string? Maybe it should be stored as a number until you actually know it will be used in JSON formatted data. 

Seems like the `modem_info_json_string_get` is anyway turning the string back to into a number while creating the JSON formatted data object.

Reply:
I've done some thinking. 

The asset tracker requires the data to be outputted to the cloud to be sent as a string. There is currently no information about the parameter type on the application level, and for the case of the asset tracker is it not desired to add any parameter type check logic either. This means there must be some way to obtain a int/short value from the modem informer as a string. 

However, it is also desirable as you describe to use the short/int values as-is, such as for battery voltage checks or some sort of temperature logic or whatever. 

I believe the best solution would be to make `modem_info_update` return the parameter value as either a string or an integer by request from the application. (Of course, requesting a string as a number value should not be allowed). My way of doing that would be to add another input argument to `modem_info_update`, which would be the requested data type, and perform logic on this argument. The input buffer would then have to change from `char` to `void`, which I am conflicted about. What are your thoughts about this solution @evenl ?

Reply:
I don't think it's the modem_info_update functions responsibility to convert the data into any other type because the modem_info_update function should not know anything about what a potential caller want. The conversion should happen in the `modem_info_json.c,` where the information of which type is needed resides. 

There is a software principle called "Single Responsibility Principle" which I think apply in this situation.

Reply:
All right, I will try a different strategy. I think the modem_info should simply have two public functions then; one for returning a parameter as a string and one for returning the parameter as a short/int value, if applicable. I will update the PR when this is done.